Issues building to Win x64 and errors in README.Win32.md

EDIT: Updated pertinent details based on conversation / investigation in comments.

Hello,

I have a need to build tcpdump for WinPE 64 environment, meaning 32-bit executables will not run. I have been following the instructions provided at https://github.com/the-tcpdump-group/tcpdump/blob/master/doc/README.Win32.md.

Steps taken on Server 2016 build server:
- install npcap
- unzip npcap SKD contents to E:\npcap_sdk 
- install cmake
- install visual studio community 2022 with C++ cmake option
- grab latest tarball of tcpdump from tcpdump.org and unpack to E:\tcpdump-4.99.4
- reboot build server
- run "x64 Native Tools Command Prompt for VS 2022" cli tool as administrator
- run: cmake "-DPCAP_ROOT=E:\npcap_sdk" -G "Visual Studio 17 2022" -A x64 E:\tcpdump-4.99.4
- run: msbuild /m /nologo /p:Configuration=Release /p:Platform=x64 tcpdump.sln

The resulting exe at Release\tcpdump.exe runs on the build system but not when copied to WinPE x64 host, which cannot run 32-bit programs. There are no errors, but if you try to run it on the command line you get nothing back, just back to the command prompt immediately.

I'm guessing it's not actually producing a 64-bit program, but don't know.

The msbuild step produces some warnings which may or may not be a problem:

"E:\tcpdump_build\tcpdump.sln" (default target) (1) ->
       "E:\tcpdump_build\netdissect.vcxproj.metaproj" (default target) (2) ->
       "E:\tcpdump_build\netdissect.vcxproj" (default target) (5) ->
       (ClCompile target) ->
         E:\tcpdump-4.99.4\addrtoname.c(156,10): warning C4996: 'gethostbyaddr': Use getnameinfo() or GetNameInfoW() in
       stead or define _WINSOCK_DEPRECATED_NO_WARNINGS to disable deprecated API warnings [E:\tcpdump_build\netdissect.
       vcxproj]
         E:\tcpdump-4.99.4\print-arista.c(146,36): warning C4244: 'function': conversion from 'uint64_t' to 'uint32_t',
        possible loss of data [E:\tcpdump_build\netdissect.vcxproj]
         E:\tcpdump-4.99.4\print-resp.c(327,5): warning C4244: '-=': conversion from '__int64' to 'int', possible loss
       of data [E:\tcpdump_build\netdissect.vcxproj]
         E:\tcpdump-4.99.4\print-resp.c(374,5): warning C4244: '-=': conversion from '__int64' to 'u_int', possible los
       s of data [E:\tcpdump_build\netdissect.vcxproj]
         E:\tcpdump-4.99.4\print-snmp.c(759,12): warning C4267: '-=': conversion from 'size_t' to 'uint32_t', possible
       loss of data [E:\tcpdump_build\netdissect.vcxproj]
         E:\tcpdump-4.99.4\print-snmp.c(1833,19): warning C4244: '=': conversion from '__int64' to 'u_int', possible lo
       ss of data [E:\tcpdump_build\netdissect.vcxproj]
         E:\tcpdump-4.99.4\print-zeromq.c(121,76): warning C4244: 'function': conversion from 'uint64_t' to 'u_int', po
       ssible loss of data [E:\tcpdump_build\netdissect.vcxproj]

    7 Warning(s)
    0 Error(s)

OK, seems that the cmake command I shared

cmake "-DPacket_ROOT=E:\npcap_sdk" -G "Visual Studio 17 2022" E:\tcpdump-4.99.4

And a follow-up msbuild command as per the docs of

msbuild /m /nologo /p:Configuration=Release tcpdump.sln

Works with a few warnings output:

         E:\tcpdump-4.99.4\addrtoname.c(156,10): warning C4996: 'gethostbyaddr': Use getnameinfo() or GetNameInfoW() in
       stead or define _WINSOCK_DEPRECATED_NO_WARNINGS to disable deprecated API warnings [E:\tcpdump_build\netdissect.
       vcxproj]
         E:\tcpdump-4.99.4\print-arista.c(146,36): warning C4244: 'function': conversion from 'uint64_t' to 'uint32_t',
        possible loss of data [E:\tcpdump_build\netdissect.vcxproj]
         E:\tcpdump-4.99.4\print-resp.c(327,5): warning C4244: '-=': conversion from '__int64' to 'int', possible loss
       of data [E:\tcpdump_build\netdissect.vcxproj]
         E:\tcpdump-4.99.4\print-resp.c(374,5): warning C4244: '-=': conversion from '__int64' to 'u_int', possible los
       s of data [E:\tcpdump_build\netdissect.vcxproj]
         E:\tcpdump-4.99.4\print-snmp.c(759,12): warning C4267: '-=': conversion from 'size_t' to 'uint32_t', possible
       loss of data [E:\tcpdump_build\netdissect.vcxproj]
         E:\tcpdump-4.99.4\print-snmp.c(1833,19): warning C4244: '=': conversion from '__int64' to 'u_int', possible lo
       ss of data [E:\tcpdump_build\netdissect.vcxproj]
         E:\tcpdump-4.99.4\print-zeromq.c(121,76): warning C4244: 'function': conversion from 'uint64_t' to 'u_int', po
       ssible loss of data [E:\tcpdump_build\netdissect.vcxproj]

But runs on the server it was built on - Server 2016 I believe - but not on WinPE x64, which can only run x64 executables.

I think I built a 32-bit instead of a 64-bit.

I tried -G "Visual Studio 17 2022 Win64" at first, but it doesn't work with that generator. My assumption is VS 2022 is 64-bit only. I suppose I can try to get VS 15 2019 on here and use "Visual Studio 15 2019 Win64"?

--- a/doc/README.Win32.md
+++ b/doc/README.Win32.md
@@ -122,7 +122,7 @@ Debug build.
 In the CMakeSettings.json tab, change cmakeCommandArgs to include
 
 ```
--DPacket_ROOT={path-to-sdk}
+-DPCAP_ROOT={path-to-sdk}
 ```
 
 where `{path-to-sdk}` is the path of the directory containing the Npcap or
@@ -147,7 +147,7 @@ Choose which configuration type to build, if you don't want the default
 Debug build.
 
 Scroll down to "Cmake variables and cache", scroll through the list
-looking for the entry for Packet_ROOT, and either type in the path of
+looking for the entry for PCAP_ROOT, and either type in the path of
 the directory containing the Npcap or WinPcap SDK or use the "Browse..."
 button to browse for that directory.
 
@@ -160,6 +160,8 @@ you can build with Build > "Build All".
 Building from the command line
 ------------------------------
 
+### Visual Studio 2017 ###
+
 Start the appropriate Native Tools command line prompt.
 
 Change to the directory into which you want to build tcpdump, possibly
@@ -169,19 +171,46 @@ of the tcpdump source directory.
 Run the command
 
 ```
-cmake "-DPacket_ROOT={path-to-sdk}" -G {generator} {path-to-tcpdump-source}
+cmake "-DPCAP_ROOT={path-to-sdk}" -G {generator} {path-to-tcpdump-source}
 ```
 
 `{path-to-sdk}` is the path of the directory containing the Npcap or
 WinPcap SDK.
 
-`{generator}` is the string "Visual Studio N YYYY", where `N` is the
-version of Visual Studio and `YYYY` is the year number for that version;
-if you are building a 64-bit version of tcpdump, `YYYY` must be followed
-by a space and "Win64".  For example, to build a 32-bit version of
-tcpdump with Visual Studio 2017, `{generator}` would be "Visual Studio
-15 2017" and to build a 64-bit version of tcpdump with Visual Studio
-2017, `{generator}` would be "Visual Studio 15 2017 Win64".
+`{generator}` is the string "Visual Studio 15 2017" to build a 32-bit
+version of tcpdump or the string "Visual Studio 15 2017 Win64" to build
+a 64-bit version of tcpdump.
+
+`{path-to-tcpdump-source}` is the pathname of the top-level source
+directory for tcpdump.
+
+Run the command
+
+```
+msbuild /m /nologo /p:Configuration={configuration} tcpdump.sln
+```
+
+where `{configuration}` can be "Release", "Debug", or "RelWithDebInfo".
+
+### Visual Studio 2019 ###
+
+Start the appropriate Native Tools command line prompt.
+
+Change to the directory into which you want to build tcpdump, possibly
+after creating it first.  One choice is to create it as a subdirectory
+of the tcpdump source directory.
+
+Run the command
+
+```
+cmake "-DPCAP_ROOT={path-to-sdk}" -G "Visual Studio 16 2019" {platform} {path-to-tcpdump-source}
+```
+
+`{path-to-sdk}` is the path of the directory containing the Npcap or
+WinPcap SDK.
+
+`{platform}` is `-A Win32` to build a 32-bit version of tcpdump or `-A
+x64` to build a 64-bit version of tcpdump.
 
 `{path-to-tcpdump-source}` is the pathname of the top-level source
 directory for tcpdump.

